### PR TITLE
Stateless remote mode: Adds passing in macaroon and tls data, and a CheckMacaroonPermissions RPC

### PR DESCRIPTION
**Pull request by orbitalturtle**
For Terminal's stateless-remote mode, as described in https://github.com/lightninglabs/lightning-terminal/issues/160, we add a few changes in this PR:
- Adds new configuration options to the lndclient, for passing in macaroon data and tls data directly in stateless remote mode.
- Also adds these features to the "basic client." For stateless-remote, the Pool server uses the basic client for connecting to LND.    
- Adds a wrapper for the CheckMacaroonPermissions RPC in LND for checking the validity of a Macaroon, so that we can use it to check validity in the Lightning Terminal. (Will open a separate PR into LND for this soon.) 

Relies on: 
https://github.com/lightningnetwork/lnd/pull/5304

### CODE REVIEW

File: lnd_services.go
Comment (guggero) [inline lnd_services.go:83]: nit: squash these white space/wording changes with the first commit so only the TLS changes are in this commit.

File: lightning_client.go
Comment (guggero) [inline lightning_client.go:3059]: I think for this we don't need an extra type (`CheckMacPermRequest`). Could just have the method signature like this:
```
CheckMacaroonPermissions(ctx context.Context, macaroon []byte, permissions []MacaroonPermission) (
		bool, error)
```

The type `MacaroonPermission` is available on the `lnd-13-0` branch (see my main comment).

File: lightning_client.go
Comment (guggero) [inline lightning_client.go:3075]: Over the RPC we should use `[]byte` (or `bytes` in proto) for the raw macaroon and also for the parameter of this method.

**Review by guggero — COMMENTED** [review]
Thank you very much for the PR (and the one in `lnd`)!

I did a first quick pass and it looks really good. Going to look at the `lnd` PR next.

Just one main thing that needs to be changed:
Since this PR requires functionality that isn't yet even merged to `lnd`, we won't be able to merge it for a while. It'll be merged to a branch `lnd-14-0` once we start working on that. For now, can you please open this against the `lnd-13-0` branch? That way you also get a new type for the `MacaroonPermission`. You can edit the base of your PR by clicking the edit button of the title. A new drop down then appears where you can switch from `master` to `lnd-13-0`.

File: lightning_client.go
Comment (orbitalturtle) [reply lightning_client.go:3075]: @guggero ah ok, so regenerate the protobufs in lnd to take bytes rather than a string?

**Comment by orbitalturtle** [comment]
Thanks for the initial review @guggero! Will get to these changes soon

**Comment by orbitalturtle** [comment]
I made those changes @guggero :)

File: lnd_services.go
Comment (guggero) [inline lnd_services.go:785]: I think a `switch{}` statement would be a bit nicer here than `if else if else`.

File: lnd_services.go
Comment (guggero) [inline lnd_services.go:815]: Unfortunately, the `SystemCertPool()` call doesn't work on Windows. But there's a workaround for using the system cert pool in gRPC, see here: https://github.com/lightningnetwork/lnd/blob/master/cmd/lncli/main.go#L101

File: basic_client.go
Comment (guggero) [inline basic_client.go:94]: Can add these two calls directly to the initialization of the `opts` variable above, then no `append()` is needed.

File: lnd_services.go
Comment (guggero) [inline lnd_services.go:772]: Any reason why this is an exported function?

File: lightning_client.go
Comment (guggero) [inline lightning_client.go:3481]: nit: remove the `0` argument.

**Review by guggero — COMMENTED** [review]
Thanks for the rebase! Just a few more minor things, this PR is getting very close.

File: lnd_services.go
Comment (orbitalturtle) [reply lnd_services.go:772]: Indeed, it comes in handy in this PR: https://github.com/lightninglabs/lightning-terminal/pull/245

File: lightning_client.go
Comment (orbitalturtle) [reply lightning_client.go:3481]: I think that argument is required unless i'm missing something, would len(permissions) be better?

**Comment by orbitalturtle** [comment]
Thanks @guggero! Made those changes

File: lightning_client.go
Comment (guggero) [inline lightning_client.go:3494]: This will only work once we pull in the version of `lnd` that has this new call. We can either wait until it's merged or temporarily add a `replace` directive to `go.mod`.

File: testdata/permissions.json
Comment (guggero) [inline testdata/permissions.json:143]: nit: This isn't indented the same way as the rest of the file.

File: lightning_client.go
Comment (guggero) [inline lightning_client.go:3483]: You're right, the second argument is only optional for maps. You have another problem here now though: this will create a slice with `len(permissions)` nil elements. And then in the loop you `append` more elements, effectively doubling the size of the slice. It's a weird behavior of Golang...

So either you have this pattern (setting the `size` to 0 but the `capacity` to the target length):

```
	rpcPermissions := make([]*lnrpc.MacaroonPermission, 0, len(permissions))
	for _, perm := range permissions {
		permission := &lnrpc.MacaroonPermission{
			Entity: perm.Entity,
			Action: perm.Action,
		}

		rpcPermissions = append(rpcPermissions, permission)
	}
```

or this one (using the index directly):

```
	rpcPermissions := make([]*lnrpc.MacaroonPermission, len(permissions))
	for idx, perm := range permissions {
		rpcPermissions[idx] = &lnrpc.MacaroonPermission{
			Entity: perm.Entity,
			Action: perm.Action,
		}
	}
```

I think both patterns are equally efficient but the second one is often preferred if you know the length beforehand.

File: macaroon_pouch.go
Comment (carlaKC) [inline macaroon_pouch.go:95]: instead of duplicating this code, we could just declare a `mac` var above this if/else and then set its value and return if it's not empty

File: lnd_services_test.go
Comment (carlaKC) [inline lnd_services_test.go:330]: Very nice! perhaps just add a godoc to this test describing the functionality we're covering.

File: lnd_services_test.go
Comment (carlaKC) [inline lnd_services_test.go:353]: could also assert on a more specific error here.

File: lnd_services.go
Comment (carlaKC) [inline lnd_services.go:101]: unnecessary change?

File: lnd_services.go
Comment (carlaKC) [inline lnd_services.go:810]: Impossible to hit this since we've selected on `tlsPath != ""`

File: basic_client.go
Comment (carlaKC) [inline basic_client.go:88]: nit: needs gofmt

**Review by carlaKC — COMMENTED** [review]
Thanks for the PR! 

Great addition, and nice to have some test coverage :)

File: lightning_client.go
Comment (orbitalturtle) [reply lightning_client.go:3494]: Ah, replace it with this branch? https://github.com/lightningnetwork/lnd/pull/5304

File: lightning_client.go
Comment (orbitalturtle) [reply lightning_client.go:3483]: got it, thanks for the explanation!

File: lnd_services.go
Comment (orbitalturtle) [reply lnd_services.go]: Ah good catch. I assume we still want to support the case where the tlsPath isn't set, but the user is relying on whatever the defaultTLSCertPath is set to. So I added that logic to the default case, if you want to check that it looks okay

File: lnd_services.go
Comment (carlaKC) [inline lnd_services.go:808]: nit: add newlines between cases?

File: lnd_services.go
Comment (carlaKC) [inline lnd_services.go:824]: Do we have a use case for this? Would imagine we can just error out if default cert isn't found at this stage.

File: lightning_client.go
Comment (carlaKC) [reply lightning_client.go:3503]: you can get the mod version using: 
```
$ go get github.com/orbitalturtle/lnd@{last commit on PR branch}
go: downloading github.com/orbitalturtle/lnd v0.0.1-beta 
```

Then this to the bottom of `go.mod`:
`replace github.com/lightningnetwork/lnd => github.com/orbitalturtle/lnd v0.0.1-beta`

**Review by carlaKC — COMMENTED** [review]
Looking good! 

Once the lnd version here is updated and the dependent loop PR points to this I'll test some of these options out.

File: lnd_services.go
Comment (orbitalturtle) [reply lnd_services.go:827]: Indeed, we sometimes need to fallback on the system pool at Voltage ~

**Comment by orbitalturtle** [comment]
Made those changes @carlaKC ! FYI I also made a few small additional changes, since we're moving in a new direction for integrated mode in the Terminal (described here https://github.com/lightninglabs/lightning-terminal/issues/160#issuecomment-888689859). In short, needed to allow the option for passing in no tls creds, so I added another boolean parameter in GetTLSCredentials to enable that option.

File: lightning_client.go
Comment (guggero) [inline lightning_client.go:3538]: nit: copy godoc from interface above.

File: go.mod
Comment (guggero) [inline go.mod:14]: Can remove the replace directive now and reference the merge commit of lnd above.

**Review by guggero — COMMENTED** [review]
Almost there, just a few more minor things. Great work pushing forward on this quite respectable chain of dependent PRs :clap:

File: lnd_services.go
Comment (orbitalturtle) [reply lnd_services.go:845]: sounds good

File: go.mod
Comment (orbitalturtle) [reply go.mod:14]: Hmmm so if I use go get github.com/lightningnetwork/lnd@5f43ef5816104f0875460ff24c697f115d97fc69 on the baking macaroon commit from #5304, I get this error:

```
github.com/lightningnetwork/lnd/lncfg
# github.com/lightningnetwork/lnd/lncfg
../../../go/pkg/mod/github.com/lightningnetwork/lnd@v0.13.0-beta.rc5.0.20210914212831-449f84d849bc/lncfg/db.go:93:43: too many arguments in call to kvdb.StartEtcdTestBackend
../../../go/pkg/mod/github.com/lightningnetwork/lnd@v0.13.0-beta.rc5.0.20210914212831-449f84d849bc/lncfg/db.go:95:37: db.Etcd.EmbeddedLogFile undefined (type *etcd.Config has no field or method EmbeddedLogFile)
../../../go/pkg/mod/github.com/lightningnetwork/lnd@v0.13.0-beta.rc5.0.20210914212831-449f84d849bc/lncfg/db.go:190:11: db.Etcd.CloneWithSubNamespace undefined (type *etcd.Config has no field or method CloneWithSubNamespace)
../../../go/pkg/mod/github.com/lightningnetwork/lnd@v0.13.0-beta.rc5.0.20210914212831-449f84d849bc/lncfg/db.go:199:11: db.Etcd.CloneWithSubNamespace undefined (type *etcd.Config has no field or method CloneWithSubNamespace)
../../../go/pkg/mod/github.com/lightningnetwork/lnd@v0.13.0-beta.rc5.0.20210914212831-449f84d849bc/lncfg/db.go:209:11: db.Etcd.CloneWithSubNamespace undefined (type *etcd.Config has no field or method CloneWithSubNamespace)
../../../go/pkg/mod/github.com/lightningnetwork/lnd@v0.13.0-beta.rc5.0.20210914212831-449f84d849bc/lncfg/db.go:219:11: db.Etcd.CloneWithSubNamespace undefined (type *etcd.Config has no field or method CloneWithSubNamespace)
../../../go/pkg/mod/github.com/lightningnetwork/lnd@v0.13.0-beta.rc5.0.20210914212831-449f84d849bc/lncfg/db.go:229:11: db.Etcd.CloneWithSubNamespace undefined (type *etcd.Config has no field or method CloneWithSubNamespace)
../../../go/pkg/mod/github.com/lightningnetwork/lnd@v0.13.0-beta.rc5.0.20210914212831-449f84d849bc/lncfg/db.go:239:11: db.Etcd.CloneWithSubNamespace undefined (type *etcd.Config has no field or method CloneWithSubNamespace)
```

But if I go get a more recent commit the error disappears. Any reason not to use the more recent commit?

Comment (guggero) [reply go.mod:14]: Ah yes, there was a problem with the `kvdb` submodule that has been fixed in the meantime. No reason against using the most recent commit from `master` :+1:

File: lightning_client.go
Comment (carlaKC) [inline lightning_client.go:3550]: Add a timeout to this `ctx` as well?

**Review by carlaKC — APPROVED** [review]
Looking good! Just need to bump deps to master to get around that `kvdb` issue. 

I think we can bump lightninglabs/loop/pull/406 to depend on this version of lndclient just to test out everything works as expected then merge.

**Review by guggero — APPROVED** [review]
LGTM :tada:

**Comment by guggero** [comment]
The linter is complaining:

```
lnd_services.go:805: G402: TLS InsecureSkipVerify set true. (gosec)
			InsecureSkipVerify: true,
```

Probably needs a ` // nolint:gosec` comment on that line.

**Comment by orbitalturtle** [comment]
@guggero made that linter change!

@carlaKC Looks like i was wrong about that loop PR requiring this lndclient change ~ will be submitting a terminal change soon that requires it though

**Pull request merged**
